diff --git a/v3-no-escape/go.mod b/v4-testing/go.mod
index a5d6f05..608c1c2 100644
--- a/v3-no-escape/go.mod
+++ b/v4-testing/go.mod
@@ -1,3 +1,3 @@
-module github.com/hammondus/webserver-evolution/ver3
+module github.com/hammondus/webserver-evolution/ver4
 
 go 1.26.3
diff --git a/v3-no-escape/index.html b/v4-testing/index.html
index a8a2033..8dd39f8 100644
--- a/v3-no-escape/index.html
+++ b/v4-testing/index.html
@@ -4,11 +4,11 @@
 <head>
     <meta charset="UTF-8">
     <meta name="viewport" content="width=device-width, initial-scale=1.0">
-    <title>Web Server v3</title>
+    <title>Web Server v4</title>
 </head>
 
 <body>
-    <h1>Web Server v3</h1>
+    <h1>Web Server v4</h1>
 </body>
 
 </html>
\ No newline at end of file
diff --git a/v3-no-escape/main.go b/v4-testing/main.go
index 1d0440b..a0dafa8 100644
--- a/v3-no-escape/main.go
+++ b/v4-testing/main.go
@@ -28,6 +28,7 @@ func main() {
 type customFileSystem struct {
 	innerFS       fs.FS
 	dotFileAccess bool
+	defaultHtml   []string
 }
 
 func NewCustomFS(rootPath string, rootEscape bool, dotFileAccess bool) (*customFileSystem, error) {
diff --git a/v4-testing/main_test.go b/v4-testing/main_test.go
new file mode 100644
index 0000000..6458318
--- /dev/null
+++ b/v4-testing/main_test.go
@@ -0,0 +1,187 @@
+package main
+
+import (
+	"net/http"
+	"net/http/httptest"
+	"testing"
+)
+
+func TestServeCustomFS_DotFilesBlocked_No_Escape(t *testing.T) {
+	fsys, err := NewCustomFS("testdata", false, false)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	r := httptest.NewRequest(http.MethodGet, "/.secretFile", nil)
+	w := httptest.NewRecorder()
+	serveCustomFS(fsys, w, r)
+	if w.Code != http.StatusNotFound {
+		t.Errorf("expected 404, got %d", w.Code)
+	}
+
+	r = httptest.NewRequest(http.MethodGet, "/.secretDir/testes.txt", nil)
+	w = httptest.NewRecorder()
+	serveCustomFS(fsys, w, r)
+	if w.Code != http.StatusNotFound {
+		t.Errorf("expected 404, got %d", w.Code)
+	}
+
+	// should get file not found when asking for a directory that doesn't contain index.html
+	r = httptest.NewRequest(http.MethodGet, "/noindex", nil)
+	w = httptest.NewRecorder()
+	serveCustomFS(fsys, w, r)
+	if w.Code != http.StatusNotFound {
+		t.Errorf("expected 404, got %d", w.Code)
+	}
+
+	// should status ok 200 when accessing a directory with an index.html
+	r = httptest.NewRequest(http.MethodGet, "/static", nil)
+	w = httptest.NewRecorder()
+	serveCustomFS(fsys, w, r)
+	if w.Code != http.StatusOK {
+		t.Errorf("expected 200, got %d", w.Code)
+	}
+
+	// shoudn't be able to esacpe to a directory at a higher level than we start
+	r = httptest.NewRequest(http.MethodGet, "/escape/README.md", nil)
+	w = httptest.NewRecorder()
+	serveCustomFS(fsys, w, r)
+	if w.Code != http.StatusNotFound {
+		t.Errorf("expected 404, got %d", w.Code)
+	}
+}
+
+func TestServeCustomFS_DotFilesBlocked_With_Escape(t *testing.T) {
+	fsys, err := NewCustomFS("testdata", true, false)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	r := httptest.NewRequest(http.MethodGet, "/.secretFile", nil)
+	w := httptest.NewRecorder()
+	serveCustomFS(fsys, w, r)
+	if w.Code != http.StatusNotFound {
+		t.Errorf("expected 404, got %d", w.Code)
+	}
+
+	r = httptest.NewRequest(http.MethodGet, "/.secretDir/testes.txt", nil)
+	w = httptest.NewRecorder()
+	serveCustomFS(fsys, w, r)
+	if w.Code != http.StatusNotFound {
+		t.Errorf("expected 404, got %d", w.Code)
+	}
+
+	// should get file not found when asking for a directory that doesn't contain index.html
+	r = httptest.NewRequest(http.MethodGet, "/noindex", nil)
+	w = httptest.NewRecorder()
+	serveCustomFS(fsys, w, r)
+	if w.Code != http.StatusNotFound {
+		t.Errorf("expected 404, got %d", w.Code)
+	}
+
+	// should status ok 200 when accessing a directory with an index.html
+	r = httptest.NewRequest(http.MethodGet, "/static", nil)
+	w = httptest.NewRecorder()
+	serveCustomFS(fsys, w, r)
+	if w.Code != http.StatusOK {
+		t.Errorf("expected 200, got %d", w.Code)
+	}
+
+	// should be able to esacpe to a directory at a higher level than we start
+	r = httptest.NewRequest(http.MethodGet, "/escape/README.md", nil)
+	w = httptest.NewRecorder()
+	serveCustomFS(fsys, w, r)
+	if w.Code != http.StatusOK {
+		t.Errorf("expected 200, got %d", w.Code)
+	}
+}
+
+func TestServeCustomFS_DotFilesEnabled_No_Escape(t *testing.T) {
+	fsys, err := NewCustomFS("testdata", false, true)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	r := httptest.NewRequest(http.MethodGet, "/.secretFile", nil)
+	w := httptest.NewRecorder()
+	serveCustomFS(fsys, w, r)
+	if w.Code != http.StatusOK {
+		t.Errorf("expected 200, got %d", w.Code)
+	}
+
+	r = httptest.NewRequest(http.MethodGet, "/.secretDir/testes.txt", nil)
+	w = httptest.NewRecorder()
+	serveCustomFS(fsys, w, r)
+	if w.Code != http.StatusOK {
+		t.Errorf("expected 200, got %d", w.Code)
+	}
+
+	// should get file not found when asking for a directory that doesn't contain index.html
+	r = httptest.NewRequest(http.MethodGet, "/noindex", nil)
+	w = httptest.NewRecorder()
+	serveCustomFS(fsys, w, r)
+	if w.Code != http.StatusNotFound {
+		t.Errorf("expected 404, got %d", w.Code)
+	}
+
+	// should status ok 200 when accessing a directory with an index.html
+	r = httptest.NewRequest(http.MethodGet, "/static", nil)
+	w = httptest.NewRecorder()
+	serveCustomFS(fsys, w, r)
+	if w.Code != http.StatusOK {
+		t.Errorf("expected 200, got %d", w.Code)
+	}
+
+	// shouldn't be able to esacpe to a directory at a higher level than we start
+	r = httptest.NewRequest(http.MethodGet, "/escape/README.md", nil)
+	w = httptest.NewRecorder()
+	serveCustomFS(fsys, w, r)
+	if w.Code != http.StatusNotFound {
+		t.Errorf("expected 404, got %d", w.Code)
+	}
+}
+
+func TestServeCustomFS_DotFilesEnabled_With_Escape(t *testing.T) {
+	fsys, err := NewCustomFS("testdata", true, true)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	r := httptest.NewRequest(http.MethodGet, "/.secretFile", nil)
+	w := httptest.NewRecorder()
+	serveCustomFS(fsys, w, r)
+	if w.Code != http.StatusOK {
+		t.Errorf("expected 200, got %d", w.Code)
+	}
+
+	r = httptest.NewRequest(http.MethodGet, "/.secretDir/testes.txt", nil)
+	w = httptest.NewRecorder()
+	serveCustomFS(fsys, w, r)
+	if w.Code != http.StatusOK {
+		t.Errorf("expected 200, got %d", w.Code)
+	}
+
+	// should get file not found when asking for a directory that doesn't contain index.html
+	r = httptest.NewRequest(http.MethodGet, "/noindex", nil)
+	w = httptest.NewRecorder()
+	serveCustomFS(fsys, w, r)
+	if w.Code != http.StatusNotFound {
+		t.Errorf("expected 404, got %d", w.Code)
+	}
+
+	// should status ok 200 when accessing a directory with an index.html
+	r = httptest.NewRequest(http.MethodGet, "/static", nil)
+	w = httptest.NewRecorder()
+	serveCustomFS(fsys, w, r)
+	if w.Code != http.StatusOK {
+		t.Errorf("expected 200, got %d", w.Code)
+	}
+
+	// shoudn be able to esacpe to a directory at a higher level than we start
+	r = httptest.NewRequest(http.MethodGet, "/escape/README.md", nil)
+	w = httptest.NewRecorder()
+	serveCustomFS(fsys, w, r)
+	if w.Code != http.StatusOK {
+		t.Errorf("expected 200, got %d", w.Code)
+	}
+}
diff --git a/v3-no-escape/.secretDir/.secretFileInSecretDir.txt b/v4-testing/testdata/.secretDir/.secretFileInSecretDir.txt
similarity index 100%
rename from v3-no-escape/.secretDir/.secretFileInSecretDir.txt
rename to v4-testing/testdata/.secretDir/.secretFileInSecretDir.txt
diff --git a/v3-no-escape/.secretDir/testes.txt b/v4-testing/testdata/.secretDir/testes.txt
similarity index 100%
rename from v3-no-escape/.secretDir/testes.txt
rename to v4-testing/testdata/.secretDir/testes.txt
diff --git a/v3-no-escape/.secretFile b/v4-testing/testdata/.secretFile
similarity index 100%
rename from v3-no-escape/.secretFile
rename to v4-testing/testdata/.secretFile
diff --git a/v3-no-escape/escape b/v4-testing/testdata/escape
similarity index 100%
rename from v3-no-escape/escape
rename to v4-testing/testdata/escape
diff --git a/v3-no-escape/noindex/anotherDirectory/anothertext.txt b/v4-testing/testdata/noindex/anotherDirectory/anothertext.txt
similarity index 100%
rename from v3-no-escape/noindex/anotherDirectory/anothertext.txt
rename to v4-testing/testdata/noindex/anotherDirectory/anothertext.txt
diff --git a/v3-no-escape/noindex/static.html b/v4-testing/testdata/noindex/static.html
similarity index 100%
rename from v3-no-escape/noindex/static.html
rename to v4-testing/testdata/noindex/static.html
diff --git a/v3-no-escape/noindex/text.txt b/v4-testing/testdata/noindex/text.txt
similarity index 100%
rename from v3-no-escape/noindex/text.txt
rename to v4-testing/testdata/noindex/text.txt
diff --git a/v3-no-escape/static/index.html b/v4-testing/testdata/static/index.html
similarity index 100%
rename from v3-no-escape/static/index.html
rename to v4-testing/testdata/static/index.html
diff --git a/v3-no-escape/static/static.html b/v4-testing/testdata/static/static.html
similarity index 100%
rename from v3-no-escape/static/static.html
rename to v4-testing/testdata/static/static.html
